src/Symfony/Component/Routing/Generator/CompiledUrlGenerator.php

```diff
@@ -40,7 +40,6 @@ public function generate($name, $parameters = [], $referenceType = self::ABSOLUT
         if (null !== $locale) {
             do {
                 if (($this->compiledRoutes[$name.'.'.$locale][1]['_canonical_route'] ?? null) === $name) {
-                    unset($parameters['_locale']);
```

Comment:
Not needed anymore because it is handled by the new code.

Reply:
wouldn't it be OK to do `$parameters['_locale'] = $locale` here, and remove the other changes?

Reply:
No because it's not enough to handle the case where the generator locale is "fr" + you ask for the "foo.en" route. In this case the locale remains "fr".